p2p/identity_service.go

```diff
@@ -0,0 +1,286 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	protocol "github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/qri-io/qri/config"
+	"github.com/qri-io/qri/event"
+	"github.com/qri-io/qri/repo"
+	"github.com/qri-io/qri/repo/profile"
+)
+
+const (
+	// ProfileProtocolID is the protocol id for the profile exchange service
+	ProfileProtocolID = protocol.ID("/qri/profile/0.1.0")
+	// ProfileTimeout is the length of time we will wait for a response in a
+	// profile exchange
+	ProfileTimeout = time.Minute * 2
+)
+
+var (
+	// expectedProtocols is the list of protocols we expect a node that has a
+	// profile service to speak
+	expectedProtocols = []string{
+		string(QriProtocolID),
+		string(ProfileProtocolID),
+	}
+	// ErrPeerNotFound is returned when the profile service cannot find the
+	// peer in question
+	ErrPeerNotFound = fmt.Errorf("peer not found")
+)
+
+// QriIdentityService manages the profile exchange. This exchange should happen
+// whenever a node connects to a new peer
+type QriIdentityService struct {
```

Comment:
would you prefer `QriProfileService` as a name, because it would be way easier to change the protocol to `/qri/identity/0.1.0`

The reason I prefer `IdentityService` is it is doing more than just exchanging profiles, it's verifying the identity of the remote peer & ensuring a lasting connection.

But this is not a hill I care about dying on, so if you prefer `QriProfileService` that's fine!

Reply:
Yeah totally agree it's doing more than profile exchange. My concern is I don't see "Identity" as being a broader concept than "Profile". We have other places within the codebase that do this profile / identity mixup, and I think it's a source of confusion.

To me the method names on this service do a really good job of articulating what it's doing. I know it's more work, but I think it's worth it.

Reply:
no prob :)